dissect/protos/dns.py

```diff
@@ -0,0 +1,575 @@
+            raise RuntimeError('DNS suspicious count threshold hit')
+        self.section.authority = DnsResourceRecordArray(self.nscount)
+
+    def pcb_arcount(self):
+        if self.arcount > DNS_SUSPICIOUS_COUNT:
+            raise RuntimeError('DNS suspicious count threshold hit')
+        self.section.additional = DnsResourceRecordArray(self.arcount)
+
+    def vsParse(self, bytez, offset=0, writeback=False):
+        self._cache_qrs = None
+        self._cache_ars = None
+        self._dns_bytes = bytez
+        self._dns_offset = offset       
+        return VStruct.vsParse(self, bytez, offset=offset)
+
+    def _getLabelPointerRef(self, msgofs):
+        '''
+        Given an offset relative to the beginning of a message, create a
+        DnsName() structure based on the data there, and return the results
+        of its getTypeVal() method (a (nametype, nameval) tuple).
+        '''
+        # msgofs is relative to the beginning of the message, not necessarily the stream
+        if not msgofs in self._nptr:
+            # these are often repeated within a message, so we cache them
+            self._nptr[msgofs] = DnsName()
+            self._nptr[msgofs].vsParse(self._dns_bytes, self._dns_offset + msgofs)
+        return self._nptr[msgofs].getTypeVal()
+
+    def getDnsName(self, nametype, nameval):
+        '''
+        Given a nametype (one of the DNS_NAMETYPE_* constants) and nameval
+        (depending on the type, either an fqdn, pointer, or partial fqdn and
+        a pointer), return an fqdn.  This is meant to be called with the
+        results from a DnsName() instance's getTypeVal() method.
+        '''
+        if nametype == DNS_NAMETYPE_LABEL:
+            fqdn = nameval
+
+        elif nametype == DNS_NAMETYPE_POINTER:
+            offset = nameval
+            if self._tcpdns:
+                # if we're a TCP packet, the 'length' field is not included in the pointer offset
+                offset += 2
+
+            fqdn = self.getDnsName(*self._getLabelPointerRef(offset))
+
+        elif nametype == DNS_NAMETYPE_LABELPOINTER:
+            beginlabels,offset = nameval
+            if self._tcpdns:
+                # if we're a TCP packet, the 'length' field is not included in the pointer offset
+                offset += 2
+
+            ptrtype,ptrval = self._getLabelPointerRef(offset)
+            endlabels = self.getDnsName(ptrtype, ptrval)
+            fqdn = '.'.join((beginlabels, endlabels))
+        return fqdn
+
+    def getQuestionRecords(self):
+        '''
+        Return a list of Question records as (dnstype, dnsclass, fqdn) tuples.
+        '''
+        if self._cache_qrs:
+            return self._cache_qrs
+        ret = []
+        for fname,q in self.section.question:
+            fqdn = self.getDnsName(*q.qname.getTypeVal())
+            ret.append((q.qtype, q.qclass, fqdn))
+        self._cache_qrs = ret
+        return ret
+
+    def _getResourceRecords(self, structure):
```

Comment:
Seems like this is a top level API ( compliment of getQuestionRecords ) lets promote him by removing the _